lib/chef/resource/macos_userdefaults.rb

```
@@ -102,149 +104,63 @@ class MacosUserDefaults < Chef::Resource
+          write_preferences(new_resource)
         end
       end
 
       action :delete, description: "Delete a key from a domain." do
         # if it's not there there's nothing to remove
         return unless current_resource
 
-        converge_by("delete domain:#{new_resource.domain} key:#{new_resource.key}") do
+        # TODO: implement in CF and use here
+      end
 
-          cmd = defaults_modify_cmd
-          Chef::Log.debug("Removing defaults key by shelling out: #{cmd.join(" ")}")
+      action_class do
+        CF_MAPPING = {
+          current_user: CF::Preferences::CURRENT_USER,
+          all_users: CF::Preferences::ALL_USERS,
+          current_host: CF::Preferences::CURRENT_HOST,
+          all_hosts: CF::Preferences::ALL_HOSTS,
+          current: CF::Preferences::CURRENT_HOST # TODO: deprecation warning for this option
+        }
+
+        def read_preferences(new_resource)
+          CF::Preferences.get(new_resource.key, new_resource.domain, mapped_user, mapped_host)
+        end
 
-          shell_out!(cmd, user: new_resource.user)
+        def write_preferences(new_resource)
+          CF::Preferences.set(new_resource.key, new_resource.value, new_resource.domain, mapped_user, mapped_host)
         end
-      end
 
-      action_class do
-        #
-        # The command used to write or delete delete values from domains
-        #
-        # @return [Array] Array representation of defaults command to run
-        #
-        def defaults_modify_cmd
-          cmd = ["/usr/bin/defaults"]
-
-          if new_resource.host == :current
-            cmd.concat(["-currentHost"])
-          elsif new_resource.host # they specified a non-nil value, which is a hostname
-            cmd.concat(["-host", new_resource.host])
-          end
-
-          cmd.concat([action.to_s, new_resource.domain, new_resource.key])
-          cmd.concat(processed_value) if action == :write
-          cmd.prepend("sudo") if new_resource.sudo
```

this may be a breaking change.  i'm not sure how it would be possible to not break it though.

---

should we put a check that if an operation is intended for some privileged domain and the `sudo` prop is false, we should raise an error/warning?